### DISCUSSION TITLE
generate JATS xml with "floating" figures and tables

### DISCUSSION THREAD

**Discussion by jhpoelen** [jgm/pandoc#10196] [Q&A]
Hi! 

I am using pandoc to create data reviews in various formats (pdf, docx, html, jats xml) for species interaction datasets indexed by Global Biotic Interactions (GloBI, https://globalbioticinteractions.org). 

And, in collaboration with a publisher, I was asked to include the figures and tables in the "back" section, instead of the current inline. It seems that by default, `pandoc` generated JATS inlines figures and tables. For context see https://github.com/globalbioticinteractions/globalbioticinteractions/issues/1014 and associated html version of the review report at https://depot.globalbioticinteractions.org/review/globalbioticinteractions/dopi 

I've included the markdown source, output jats xml, and the command to connect them below

input: 
[index.md](https://github.com/user-attachments/files/17050260/index.md)

output:
[index.xml.txt](https://github.com/user-attachments/files/17050278/index.xml.txt)

```bash
cat index.md\
 | pandoc --embed-resources --standalone --toc --citeproc -t jats -o -\
 > index.xml.txt
```

for context see also https://github.com/globalbioticinteractions/globinizer/blob/52c0dec75ba425b8bac941e32659d7568080e7fc/check-dataset.sh#L781C3-L781C68 .

Curious to hear your thoughts on how to generate JATS with "floating" figures and tables . . . 

fyi @Daniel-Mietchen @teodorgeorgiev

**Comment by jgm** [comment]
@tarleb do you have thoughts on this?

**Reply by tarleb** [reply]
I think this would be a reasonable feature request: the *publishing* and *archiving* tag sets support this usage, but the *articleauthoring* tag set does not.

The spec says that the `<float-group>` element must be in the `<article>`, not the back-matter, so there might be a mistake in the linked issue.

**Reply by jhpoelen** [reply]
@jgm @tarleb thanks for your prompt reply and for sharing your observations.

Please let me know if there's anything I can do to help move this forward. Unfortunately, my Haskell coding stills are non-existing, so creating a pull request may take some time and significant investment in my end (e.g., learning Haskell).

**Reply by estedeahora** [reply]
if this change is advanced, could it be an optional behavior?

**Reply by teodorgeorgiev** [reply]
indeed **float-group** should be in **article** (after **back**, not in **back**)

**Reply by tarleb** [reply]
I'm having second thoughts about adding this to pandoc proper, so here's a [custom Lua writer](https://pandoc.org/custom-writers) to produce the modified JATS. It's a tad more complicated than I expected, as JATS templates use partials, which made modifying the right part of the template more difficult.

Usage: `pandoc --to=jats-float-group.lua ...`. Please let me know if it works as intended.

``` lua
--- File: jats-float-group.lua
--- Author: Albert Krewinkel
--- License: MIT

local path     = require 'pandoc.path'
local system   = require 'pandoc.system'
local template = require 'pandoc.template'

local target_format = 'jats_publishing'

Extensions = pandoc.format.extensions(target_format)
Template = pandoc.template.default(target_format)

local function write_file(filepath, content)
  local fh = io.open(filepath, 'w')
  fh:write(content)
  fh:close()
end

function Writer (doc, opts)
  local format = {
    format = target_format,
    extensions = opts.extensions
  }

  -- Collect floats in a separate list, remove from main doc.
  local floats = pandoc.Blocks{}
  local move_to_floats = function (x)
    floats:insert(x)
    return {}
  end
  doc = doc:walk {
    Figure = move_to_floats,
    Table = move_to_floats,
  }

  -- Store floats in the metadata
  doc.meta['floats-group'] = floats

  -- modify the template on the fly, add `<floats-group>`
  if opts.template then
    local article = template.get 'article.jats_publishing'
    article = article:gsub(
      '</article>',
      '<floats-group>\n${floats-group}\n</floats-group>\n</article>'
    )
    local affiliations = template.get 'affiliations.jats'
    local deftemplate = template.default(target_format)

    system.with_temporary_directory('jats-template', function (tmpdir)
      local template_path = path.join{tmpdir, 'default.jats_publishing'}
      write_file(path.join{tmpdir, 'article.jats_publishing'}, article)
      write_file(path.join{tmpdir, 'affiliations.jats'}, affiliations)
      write_file(template_path, deftemplate)
      opts.template = template.compile(
        template.default(target_format),
        template_path)
    end)
  end

  return pandoc.write(doc, format, opts)
end
```

**Reply by denismaier** [reply]
> if this change is advanced, could it be an optional behavior?

Yes, I also think it should be optional.

**Reply by tarleb** [reply]
I'm opting for a simple change that will simply add support for a `floats-group` variable to the template. The user then has the option to set the value manually in the metadata, or to use a filter.

E.g.

``` yaml
---
floats-group:
  - |
    ![figure](image.png)

  - |
    | table |
    |--
    | example |
---
```

The above Lua filter to set this value automatically becomes considerably simpler:

``` lua
--- jats-floats-group.lua -- output JATS with figures, tables in separate tag
--- Author: Albert Krewinkel
--- License: MIT

local pandoc   = require 'pandoc'
local target_format = 'jats_publishing'

Extensions = pandoc.format.extensions(target_format)
Template = pandoc.template.default(target_format)

function Writer (doc, opts)
  local format = {
    format = target_format,
    extensions = opts.extensions
  }

  -- Collect floats in a separate list, remove from main doc.
  local floats = pandoc.Blocks{}
  local move_to_floats = function (x)
    floats:insert(x)
    return {}
  end
  doc = doc:walk {
    Figure = move_to_floats,
    Table = move_to_floats,
  }

  -- Store floats in the metadata
  doc.meta['floats-group'] = floats

  return pandoc.write(doc, format, opts)
end
```

Commit incoming.

**Comment by jhpoelen** [comment]
fyi @myrmoteras
